### ISSUE TITLE
ecma(javascript) injection has massive performance hits when $(`...`) is present

### ISSUE THREAD

**Issue by Hippo0o** [nvim-treesitter/nvim-treesitter#5316]
### Describe the bug

The parser experiences significant slowdown when a JavaScript file contains a function call with a $ sign and a template string as its parameter.

I could find the line responsible at https://github.com/nvim-treesitter/nvim-treesitter/blob/4d41d9bfb09dd0836ce6910404e3cd570500c9ca/queries/ecma/injections.scm#L10

I'm not sure if it's necessary to consider every function as a potential injection target, especially when some functions are already explicitly defined.

### To Reproduce

1. open some js file with ```$(`...`)``` in it
2. start typing in insert mode
3. massive lag

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- WARNING `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
- OK `node` found v20.5.1 (only needed for :TSInstallFromGrammar)
- OK `git` executable found.
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: cc (GCC) 13.2.1 20230801
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "x86_64",
  release = "6.4.10-zen2-1-zen",
  sysname = "Linux",
  version = "#1 ZEN SMP PREEMPT_DYNAMIC Sun, 13 Aug 2023 01:33:59 +0000"
} ~

Parser/Features         H L F I J
  - bash                ✓ ✓ ✓ . ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - css                 ✓ . ✓ ✓ ✓
  - dockerfile          ✓ . . . ✓
  - html                ✓ ✓ ✓ ✓ ✓
  - http                ✓ . . . ✓
  - javascript          ✓ ✓ ✓ ✓ ✓
  - json                ✓ ✓ ✓ ✓ .
  - lua                 ✓ ✓ ✓ ✓ ✓
  - markdown            ✓ . ✓ ✓ ✓
  - markdown_inline     ✓ . . . ✓
  - php                 ✓ ✓ ✓ ✓ ✓
  - phpdoc              ✓ . . . .
  - python              ✓ ✓ ✓ ✓ ✓
  - query               ✓ ✓ ✓ ✓ ✓
  - scss                ✓ . ✓ ✓ .
  - sql                 ✓ . . ✓ ✓
  - twig                ✓ . . . ✓
  - typescript          ✓ ✓ ✓ ✓ ✓
  - vim                 ✓ ✓ ✓ . ✓
  - vue                 ✓ . ✓ ✓ ✓
  - yaml                ✓ ✓ ✓ ✓ ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~
```

### Output of `nvim --version`

```text
NVIM v0.10.0-dev-948+g0b0d91276
Build type: RelWithDebInfo
LuaJIT 2.1.1692716794
Run "nvim -V1 -v" for more info
```

**Comment by amaanq** [comment]
This will sound dumb, but try changing it to this:

`function: (identifier) @injection.language`

And a file that reproduces this would be great.

**Comment by Hippo0o** [comment]
@amaanq just a simple file like `test.js`:
```javascript
$(`test`)
```
i also tried the change you suggested but there is no difference.

**Comment by Hippo0o** [comment]
i should clarify that its most noticeable in insert mode when typing

**Comment by amaanq** [comment]
I can reproduce this, yeah that's really laggy

**Comment by amaanq** [comment]
Does adding this above the offset! directive help?

```scm
(call_expression
 function: ((identifier) @injection.language)
 arguments: [
             (arguments
              (template_string) @injection.content)
             (template_string) @injection.content
            ]
     (#lua-match? @injection.language "[a-zA-Z][a-zA-Z0-9]*")
     (#offset! @injection.content 0 1 0 -1)
     (#not-eq? @injection.language "svg"))
```

**Comment by Hippo0o** [comment]
@amaanq it would need something like this:
```
(#lua-match? @injection.language "^[a-zA-Z][a-zA-Z0-9]*$")
``` 
Because with your example the problem still persists when the function name is something like `$test` or `te$t`.

I also wanna note that the `_` is allowed in function names. But that probably has no relevance when matching for languages.

**Comment by lucario387** [comment]
tl;dr: we need to narrow more queries :D

**Comment by amaanq** [comment]
> @amaanq it would need something like this:
> 
> ```
> (#lua-match? @injection.language "^[a-zA-Z][a-zA-Z0-9]*$")
> ```
> 
> Because with your example the problem still persists when the function name is something like `$test` or `te$t`.
> 
> I also wanna note that the `_` is allowed in function names. But that probably has no relevance when matching for languages.

Totally fine, feel free to submit a PR (I completely forgot about this)
